File: problems/026-bidirectional-pipes/solutions/python/main.py
import sys

it = iter(sys.stdin.buffer.read().split())
C = int(next(it))
n = [int(next(it)), int(next(it))]
a = [[], []]
for w in range(2):
    for _ in range(n[w]):
        t = next(it)
        a[w].append((t, 0 if t == b"C" else int(next(it))))
pc = [0, 0]
closed = [not n[0], not n[1]]
occ = [0, 0]
steps = 0
while True:
    if pc == n:
        print("SUCCESS", steps, *occ)
        break
    progress = False
    for w in range(2):
        if pc[w] == n[w]:
            continue
        t, k = a[w][pc[w]]
        o = 1 - w
        z = 0
        if t == b"W":
            if C - occ[w] >= k:
                occ[w] += k
                z = 1
        elif t == b"R":
            if occ[o] >= k:
                occ[o] -= k
                z = 1
            elif closed[o]:
                z = -1
        else:
            closed[w] = True
            z = 1
        if z < 0:
            print("FAIL", "AB"[w], steps, *occ)
            raise SystemExit
        if z:
            pc[w] += 1
            steps += 1
            progress = True
            closed[w] |= pc[w] == n[w]
    if not progress:
        print("DEADLOCK", steps, *occ)
        break

File: problems/026-bidirectional-pipes/solutions/rust/main.rs
use std::io::{self, Read};
#[derive(Clone, Copy)]
struct A {
    t: u8,
    k: i64,
}
fn main() {
    let mut s = String::new();
    io::stdin().read_to_string(&mut s).unwrap();
    let mut it = s.split_whitespace();
    let c: i64 = it.next().unwrap().parse().unwrap();
    let n = [
        it.next().unwrap().parse::<usize>().unwrap(),
        it.next().unwrap().parse::<usize>().unwrap(),
    ];
    let mut a = [Vec::new(), Vec::new()];
    for w in 0..2 {
        for _ in 0..n[w] {
            let t = it.next().unwrap().as_bytes()[0];
            let k = if t == b'C' {
                0
            } else {
                it.next().unwrap().parse().unwrap()
            };
            a[w].push(A { t, k })
        }
    }
    let (mut pc, mut closed, mut occ, mut steps) =
        ([0usize; 2], [n[0] == 0, n[1] == 0], [0i64; 2], 0i64);
    loop {
        if pc == n {
            println!("SUCCESS {steps} {} {}", occ[0], occ[1]);
            break;
        }
        let mut progress = false;
        for w in 0..2 {
            if pc[w] == n[w] {
                continue;
            }
            let x = a[w][pc[w]];
            let o = 1 - w;
            let z = if x.t == b'W' {
                if c - occ[w] < x.k {
                    0
                } else {
                    occ[w] += x.k;
                    1
                }
            } else if x.t == b'R' {
                if occ[o] < x.k {
                    if closed[o] { -1 } else { 0 }
                } else {
                    occ[o] -= x.k;
                    1
                }
            } else {
                closed[w] = true;
                1
            };
            if z < 0 {
                println!(
                    "FAIL {} {steps} {} {}",
                    if w == 0 { "A" } else { "B" },
                    occ[0],
                    occ[1]
                );
                return;
            }
            if z == 1 {
                pc[w] += 1;
                steps += 1;
                progress = true;
                if pc[w] == n[w] {
                    closed[w] = true
                }
            }
        }
        if !progress {
            println!("DEADLOCK {steps} {} {}", occ[0], occ[1]);
            break;
        }
    }
}
